### DISCUSSION TITLE
Can SU2 output the thermal conductivity field and diffusion heat flux on the wall?

### DISCUSSION THREAD

**Discussion by LDK29** [su2code/SU2#1888] [Q&A] [answered]
Hello all,
I am new to SU2. And I want to use it to do a simulation of a hypersonic cylinder flow with a catalytic wall boundary condition. I can only get the total wall heat flux now, but the diffusion heat flux is also important. So anyone knows how to output the diffusion heat flux? And if possible, I also want to output the thermal conductivity, diffusion flux, and total enthalpy fields.
Any help will be appreciated.

**Comment by bigfooted** [comment, marked as answer]
Until recently the thermal conductivity could only be given as a constant value, so it is not written in the output. If your conductivity is constant, you can still compute the thermal flux. Total enthalpy can be obtained from the energy field. Maybe some high speed experts like @WallyMaier or @jtneedels can correct and add to my understanding.

If your thermal conductivity is not constant, please make a feature request to write this as a (paraview) output.

**Reply by pcarruscag** [reply]
Heat flux is an output. A wall only has "diffusion" heat flux, there is no convection across a wall.

**Reply by LDK29** [reply]
> Until recently the thermal conductivity could only be given as a constant value, so it is not written in the output. If your conductivity is constant, you can still compute the thermal flux. Total enthalpy can be obtained from the energy field. Maybe some high speed experts like @WallyMaier or @jtneedels can correct and add to my understanding.
> 
> If your thermal conductivity is not constant, please make a feature request to write this as a (paraview) output.
Many thanks for your advice. Yes, my conductivity is not a constant. I want to simulate a thermochemical non-equilibrium flow and I will make a feature request later.

**Reply by LDK29** [reply]
Hi Pedro
Thanks for your reply.
I mean an isothermal full catalytic wall boundary condition. As far as I know,  in thermochemical non-equilibrium flow, the wall heat flux usually contains three terms: diffusion heat transfer, translational temperature heat transfer, and vibrational temperature heat transfer. My objective is to get all these terms respectively.

**Reply by pcarruscag** [reply]
Ah got it. We don't have those terms separate, you can have a look at CFVMFlowSolverBase.inl around line 2592, where the 3 terms you mention are summed.

**Reply by WallyMaier** [reply]
Hi, 
I have in several branches included the thermal conductivities for both temperatures. I'll go ahead and make a PR.    

As far as specific heat fluxes, those are not computed/saved separately

**Reply by LDK29** [reply]
Hi Wally and Pedro,
Thanks for your help.
I think if I can get the thermal conductivities, the convection terms can be calculated as the production of normal temperature gradient and thermal conductivities.

**Reply by jtneedels** [reply]
Hi,
Yes, just to confirm the total heat flux is computed in CFVMFlowSolverBase.inl as Pedro mentions, consisting of the translational temperature HF, the vibrational temperature HF, and the diffusion HF. The output currently is the summed value. You could pull the values directly from there, or modify the CNEMOCompOutput.cpp file to add specific output options that would you like. As Wally mentions, we have a feature branch that allows output of the thermal conductivity at each cell, but it's currently not in master.

**Reply by LDK29** [reply]
Hi, thanks for your interpretations
As your said, I have checked the CFVMFlowSolverBase.inl and CNEMOCompOutput.cpp. It is found that the function AddVolumeOutput() needs 4 inputs as I list below:
*\param[in] name - Name for referencing it (in the config file and in the code).
 *\param[in] field_name - Header that is printed in the output files.
 *\param[in] groupname - The name of the group this field belongs to.
  *\param[in] description - Description of the volume field.
I'm confused that the HF in CFVMFlowSolverBase.inl is "HeatFlux" but in CNEMOCompOutput.cpp is output by: 
![out](https://user-images.githubusercontent.com/113524507/212520105-24e2ac6d-db90-45d9-81a7-83b40ed06d8e.JPG)

And to output the diffusion HF, I think I need to follow the steps:
1. Find the variable name of diffusion HF in CFVMFlowSolverBase.inl.  ( I find it now,  it's "sumJhs" at about line 2613. I think the sumJhs is an array representing a field, not just a value, right? )
![sumJhs](https://user-images.githubusercontent.com/113524507/212520015-0573f28d-689d-46a8-87c6-d94b10e678b3.JPG)

2. Add it to the CNEMOCompOutput.cpp file via the function AddVolumeOutput(). (The inputs may be "sumJhs, field name, PRIMITIVE, description"?)
3. Rebuild the src using meson.build.
Is that right? @jtneedels

**Reply by jtneedels** [reply]
Hi @LDK29,

The general outline you have is correct, but the specific steps need to be modified. 

Yes, you are correct that the sumJhs terms is the surface heat flux due to mass diffusion (catalytic effects). So if you cout that variable, you can directly access the values. 

To output a quantity in CNEMOCompOutput.cpp, you also need to fill in the SetVolumeOutput method in the function CNEMOCompOutput::LoadVolumeData. Here the necessary data is accessed from the correct place in the code. Currently only the total heat flux (the volume output you mention) is stored to be output. The heat flux associated with Ttr, Tve, and sumJhs are not. As you can see in the script above, these are created as local variables and added to HeatFlux[iMarker][iVertex]. Since they are not stored anywhere else, they can be accessed in CNEMOCompOutput.

So at present, to reconstruct the heat flux, you would need to output each piece of data individually stored: for Ttr & Tve this is the appropriate thermal conductivity value and the appropriate wall normal temperature gradient (as @WallyMaier replied, there is a PR to add these to the outputs, see #1889 ). For the catalytic flux you would need to add volume outputs for each of the components of the enthalpy diffusion flux stored in the variable class (nodes) : Grad_PrimVar, Ds, hs, etc., in the code block above. From multiplying those values you would get the component of HF due to catalytic effects.

Thinking about this some more, the easiest option is probably just to use the code from #1889 to compute k_tr * dTtr/dn and k_ve * dTve/dn, which will give you the HF due to translational and vibrational modes, then subtract that from the total HF to get the contribution of sumJhs, i.e.

HeatFlux =  k_tr * dTtr/dn + k_ve * dTve/dn + sumJhs => HeatFlux -  k_tr * dTtr/dn - k_ve * dTve/dn = sumJhs,

from which you can recover all three contributions to the heat flux.

This is a good note though, and per @bigfooted I'll add a feature request to break out reporting of heat flux due to the three components above for the NEMO solver.

**Reply by LDK29** [reply]
Hi,@jtneedels 
Thanks for your information. 
When I send the outline for confirmation I haven't seen the code 1889 yet. Because I am not an experienced user of github. I will follow your suggestion to get heat fluxes.
